crates/subspace-farmer/src/utils/farmer_piece_getter.rs

```diff
@@ -1,29 +1,50 @@
+use crate::utils::archival_storage_info::ArchivalStorageInfo;
 use crate::utils::piece_cache::PieceCache;
 use async_trait::async_trait;
+use std::collections::HashSet;
 use std::error::Error;
 use std::sync::Arc;
 use subspace_core_primitives::{Piece, PieceIndex};
 use subspace_farmer_components::plotting::{PieceGetter, PieceGetterRetryPolicy};
+use subspace_networking::libp2p::PeerId;
 use subspace_networking::utils::multihash::ToMultihash;
+use subspace_networking::utils::piece_provider::{PieceProvider, PieceValidator, RetryPolicy};
+use subspace_networking::Node;
 
-pub struct FarmerPieceGetter<PG, PC> {
-    base_piece_getter: PG,
+pub struct FarmerPieceGetter<PV, PC> {
+    node: Node,
+    piece_provider: PieceProvider<PV>,
     piece_cache: Arc<tokio::sync::Mutex<PC>>,
+    archival_storage_info: ArchivalStorageInfo,
 }
 
-impl<PG, PC> FarmerPieceGetter<PG, PC> {
-    pub fn new(base_piece_getter: PG, piece_cache: Arc<tokio::sync::Mutex<PC>>) -> Self {
+impl<PV, PC> FarmerPieceGetter<PV, PC> {
+    pub fn new(
+        node: Node,
+        piece_provider: PieceProvider<PV>,
+        piece_cache: Arc<tokio::sync::Mutex<PC>>,
+        archival_storage_info: ArchivalStorageInfo,
+    ) -> Self {
         Self {
-            base_piece_getter,
+            node,
+            piece_provider,
             piece_cache,
+            archival_storage_info,
+        }
+    }
+
+    fn convert_retry_policy(retry_policy: PieceGetterRetryPolicy) -> RetryPolicy {
+        match retry_policy {
+            PieceGetterRetryPolicy::Limited(retries) => RetryPolicy::Limited(retries),
+            PieceGetterRetryPolicy::Unlimited => RetryPolicy::Unlimited,
         }
     }
```

Comment:
I think what you want instead is `From` implementation between two types. Also I don't quite see without opening in IDE why two types are necessary.